aQute.libg/src/aQute/libg/uri/URIUtil.java

```diff
@@ -41,4 +44,83 @@ public static URI resolve(URI baseURI, String reference) throws URISyntaxExcepti
 		return baseURI.resolve(refURI);
 	}
 
+	private static final URI EMPTYURI = URI.create("");
+
+	/**
+	 * Attempts to fetch a path on the file system for the given URI string.
+	 * Tries a few more tricks than the standard method of
+	 * <tt>Paths.get(new URI(uriString))</tt> - it can handle plain paths, and
+	 * also nested URI pseudo-schemes like <tt>reference:</tt> and
+	 * <tt>jar:</tt>. Examples:
+	 * <ul>
+	 * <li>/path/to/file => /path/to/file</li>
+	 * <li>reference:file:/path/to/file => /path/to/file</li>
+	 * <li>jar:file:/path/to/file.jar!/some/contained/element =>
+	 * /path/to/file.jar</li>
+	 * <li>http://server/path/to/file => <tt>null</tt></li>
+	 * </ul>
+	 *
+	 * @param uriString The string containing the URI for which we are
+	 *            attempting to construct a path.
+	 * @return The method's best guess as to which file on the local filesystem
+	 *         this URI refers, or <tt>null</tt> if it's invalid or not a local
+	 *         filesystem URI.
+	 */
+	public static Path fromURI(String uriString) {
```

Comment:
I think the method name `fromURI` is not descriptive enough. Perhaps `pathFromURI` is more appropriate. Then its use reads better: `Path p = URIUtil.pathFromURI(loc);` or `Path p = pathFromURI(loc);` if you statically import the method.

Reply:
Originally I thought that the path prefix would be redundant, as in an expression like `Path p = fromURI(loc)` it is clear from the LHS that it's a path. However, this would not be clear in all contexts so I'll put the path prefix in.